src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesProducer.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.codec.KNN80Codec;
+
+import lombok.NonNull;
+import lombok.extern.log4j.Log4j2;
+import org.apache.lucene.codecs.DocValuesProducer;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.SegmentReadState;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FSDirectory;
+import org.apache.lucene.store.FilterDirectory;
+import org.opensearch.common.io.PathUtils;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.index.codec.util.KNNCodecUtil;
+import org.opensearch.knn.index.engine.KNNEngine;
+import org.opensearch.knn.index.memory.NativeMemoryCacheManager;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.opensearch.knn.common.KNNConstants.MODEL_ID;
+import static org.opensearch.knn.index.mapper.KNNVectorFieldMapper.KNN_FIELD;
+
+@Log4j2
+public class KNN80DocValuesProducer extends DocValuesProducer {
+
+    private final SegmentReadState state;
+    private final DocValuesProducer delegate;
+    private final NativeMemoryCacheManager nativeMemoryCacheManager;
+    private final Map<String, String> indexPathMap = new HashMap();
+
+    public KNN80DocValuesProducer(DocValuesProducer delegate, SegmentReadState state) {
+        this.delegate = delegate;
+        this.state = state;
+        this.nativeMemoryCacheManager = NativeMemoryCacheManager.getInstance();
+
+        Directory directory = state.directory;
+        // directory would be CompoundDirectory, we need get directory firstly and then unwrap
+        if (state.directory instanceof KNN80CompoundDirectory) {
+            directory = ((KNN80CompoundDirectory) state.directory).getDir();
+        }
+
+        FSDirectory fsDirectory = null;
+        try {
+            fsDirectory = ((FSDirectory) FilterDirectory.unwrap(directory));
+        } catch (ClassCastException ex) {
+            log.warn("{} can not casting to FSDirectory", directory.toString());
+            return;
+        }
```

Comment:
can we change this behavior of class casting to say instance of check.
Directory dir = FilterDirectory.unwrap(directory))
if ((dir instanceof FSDirectory) == false) {
   log.warn("{} can not casting to FSDirectory", directory);
}

Reply:
LGTM

Reply:
FIXED at https://github.com/opensearch-project/k-NN/pull/1946/commits/3b5751c08732d0a147880627fd44ae8b3196fe98